pkg/power/accelerator/source/gpu_nvml.go

```diff
@@ -0,0 +1,123 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"k8s.io/klog/v2"
+)
+
+var (
+	// List of GPU identifiers for the device
+	devices []nvml.Device
+)
+
+type GPUNvml struct {
+	collectionSupported bool
+}
+
+// Init initizalize and start the GPU metric collector
+// the nvml only works if the container has support to GPU, e.g., it is using nvidia-docker2
+// otherwise it will fail to load the libnvidia-ml.so.1
+func (n *GPUNvml) Init() (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("could not init nvml: %v", err)
+			klog.Infoln(err)
+		}
+	}()
+	if ret := nvml.Init(); ret != nvml.SUCCESS {
+		n.collectionSupported = false
+		err = fmt.Errorf("failed to init nvml: %v", nvml.ErrorString(ret))
+		return err
+	}
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		nvml.Shutdown()
+		n.collectionSupported = false
+		err = fmt.Errorf("failed to get nvml device count: %v", nvml.ErrorString(ret))
+		return err
+	}
+	klog.V(1).Infof("found %d gpu devices\n", count)
+	devices = make([]nvml.Device, count)
+	for i := 0; i < count; i++ {
+		device, ret := nvml.DeviceGetHandleByIndex(i)
+		if ret != nvml.SUCCESS {
+			nvml.Shutdown()
+			n.collectionSupported = false
+			err = fmt.Errorf("failed to get nvml device %d: %v ", i, nvml.ErrorString(ret))
+			return err
+		}
+		devices[i] = device
+	}
+	return nil
+}
+
+// Shutdown stops the GPU metric collector
+func (n *GPUNvml) Shutdown() bool {
+	return nvml.Shutdown() == nvml.SUCCESS
+}
+
+// GetGpus returns a map with gpu device
+func (n *GPUNvml) GetGpus() []nvml.Device {
+	return devices
+}
+
+// GetGpuEnergyPerGPU returns a map with mJ in each gpu device
+func (n *GPUNvml) GetGpuEnergyPerGPU() []uint32 {
+	gpuEnergy := []uint32{}
+	for _, device := range devices {
+		power, ret := device.GetPowerUsage()
+		if ret != nvml.SUCCESS {
+			klog.V(2).Infof("failed to get power usage on device %v: %v\n", device, nvml.ErrorString(ret))
+			continue
+		}
+		gpuEnergy = append(gpuEnergy, power)
+	}
+	return gpuEnergy
+}
+
+// GetProcessResourceUtilization returns a map of ProcessUtilizationSample where the key is the process pid
+//
+//	ProcessUtilizationSample.SmUtil represents the process Streaming Multiprocessors - SM (3D/Compute) utilization in percentage.
+//	ProcessUtilizationSample.MemUtil represents the process Frame Buffer Memory utilization Value.
+func (n *GPUNvml) GetProcessResourceUtilizationPerDevice(device nvml.Device, since time.Duration) (map[uint32]nvml.ProcessUtilizationSample, error) {
+	processAcceleratorMetrics := map[uint32]nvml.ProcessUtilizationSample{}
+	lastUtilizationTimestamp := uint64(time.Now().Add(-1*since).UnixNano() / 1000)
+
+	processUtilizationSample, ret := device.GetProcessUtilization(lastUtilizationTimestamp)
```

Comment:
This is cool! Process utilization is a better proxy of energy consumption. 

However, per nvidia doc[https://docs.nvidia.com/deploy/nvml-api/group__nvmlGridQueries.html#group__nvmlGridQueries_1gb0ea5236f5e69e63bf53684a11c233bd], this function is only supported in the GRID GPU. The GRID GPU product sheet is [here](https://www.nvidia.com/content/grid/virtualization-guide.pdf), I don't see anything I have. Do you have a regular GPU to test it?

Reply:
Per nvidia doc again, [nvmlDeviceGetAccountingStats](https://docs.nvidia.com/deploy/nvml-api/group__nvmlAccountingStats.html#group__nvmlAccountingStats_1g105072bad13b3cdcf0f5d20c8004aee3) is a better API for that. But note, the [gpuUtilization](https://docs.nvidia.com/deploy/nvml-api/structnvmlAccountingStats__t.html#structnvmlAccountingStats__t_1a4f64a27072be2e1f4fcc408c8a87faf) stats is for lifetime, not per sample.

Reply:
To implement this, I was inspired in the [nvtop](https://github.com/Syllo/nvtop/blob/master/src/extract_gpuinfo_nvidia.c#L567), which use this function to get the process GPU utilization.
As far as I understood, they can get the process GPU utilization without enabling NVIDIA GRID (i.e., enabling virtual GPUs - vGPUs).

Additionally, I thought that it could be possible to access  the process GPU utilization because there are two structures: `ProcessUtilizationSample` and `VgpuProcessUtilizationSample`. So that, the processes using the GRID GPUs (the vGPUs) will have the info in the `VgpuProcessUtilizationSample` and the processes directly accessing the physical gpu would have the info in the `ProcessUtilizationSample`.

The doc says that the function `GetProcessUtilization` is supported for Maxwell or newer GPUs, so most GPUs might support it.

But you are right, we need to test it in a real GPU. I will try to access it.

Reply:
At first I was thinking of using `GetAccountingStats`.
However, it only has information after the process has finished.

As we must have the information during the execution of the process, we cannot use it...

Reply:
sounds good. As long as this works on conventional GPU, it is good to me.